### PR TITLE
Fix burning mobs and isOnFire() method

### LINKED ISSUE #625: If you instant kill a mob with fire aspect enchant tool, it will not give fire aspect drops
# 🐞 I found a bug

### 📸 Screenshots / Videos

Screenshot

### ▶ Steps to Reproduce

1. Make mob health almost near to 0 for instant kill.
2. Kill it with fire aspect weapon.
3. It will not drop fire aspect drop like cooked chicken.

### ✔ Expected Behavior

It will give fire aspect drops like cooked chicken.

### ❌ Actual Behavior

It will give normal drops like raw chicken.

### 📋 Debug information

* PowerNukkit version: 1.4.0.0-SNAPSHOT (6966b19b2d104d883e319403312c5e33321b0b13)
* Debug link: -
* Timings link (if relevant): -

### 💢 Crash Dump, Stack Trace and Other Files

It doesn't crash.

### 💬 Anything else we should know?

https://github.com/CloudburstMC/Nukkit/issues/1453

### PR DESCRIPTION
**Pull request by Jl1mbo**
Fix #625

### CODE REVIEW

**Review by joserobjr — CHANGES_REQUESTED** [review]
Thank you for the pr.

Please, instead of making the post attack execute before attack, make a new method in enchantments called doPreAttack and make that execute in the same way as doPostAttack, but before, and keep the post attack after the attack. Also move the enchantment code that sets on fire after the attack to the new method. I think this would be a better solution.

Now for the isOnFire change, I'm not sure if that is right because as far as I know, we can set some fire resistant entities on fire, and they will render burning but won't take damage. However I'm speaking from my memory, I'm going to do some tests when I wake up.

**Comment by AntonPPDS** [comment]
IMHO this isn't a good way to fix. It is better if the post attacks were applied after the damage event is completed, since this is the main event, with the cancellation of which post attacks are also expected to be canceled.

**Comment by roridev** [comment]
> IMHO this isn't a good way to fix. It is better if the post attacks were applied after the damage event is completed, since this is the main event, with the cancellation of which post attacks are also expected to be canceled.

Agreed, damage cancellation should cancel all of it.   
As it stands the order of operations is confused and the behaviour is not what is is expected, you shouldn't need to cancel fire damage alone when you cancel *all* damage. It should at least back-propagate, or even better, use an sensible order of operations.

**Comment by joserobjr** [comment]
Yes, you are right. I got a new more involving idea. I'm going to change the code and ask your reviews to know what you think :)

**Comment by joserobjr** [comment]
@AntonPPDS @roridev what do you think of the new changes now?

**Review by joserobjr — APPROVED** [review]

File: src/main/java/cn/nukkit/event/entity/EntityDamageEvent.java [hunk 1]
```diff
@@ -109,6 +116,56 @@ public void setAttackCooldown(int attackCooldown) {
...
+    @PowerNukkitOnly
+    @Since("FUTURE")
+    public void setSideEffects(@Nonnull Collection<SideEffect> sideEffects) {
+        this.setSideEffects(sideEffects.toArray(SideEffect.EMPTY_ARRAY));
+    }
+
+    @PowerNukkitOnly
+    @Since("FUTURE")
+    public void addSideEffects(@Nonnull SideEffect... sideEffects) {
+        this.sideEffects = Stream
```
Comment (roridev) [inline src/main/java/cn/nukkit/event/entity/EntityDamageEvent.java:152]: Isn't this just a very verbose `Streams.concat` ?
like, `flatmapping` into `id` is just concatenation.
you're taking an `[[a]]` and converting it into an `[a]`.

```
:t concat
concat :: Foldable t => t [a] -> [a]
```

thats already implemented in the java streams api under the pure static function `Streams.concat`.

the code will end up looking like this:
```java
Stream<SideEffect> safeInput = Arrays.stream(sideEffects) // tail
                .filter(Objects::nonNull)
                .map(SideEffect::clone);

this.sideEffects = Stream.concat(Arrays.stream(this.sideEffects), safeInput).toArray(SideEffect[]::new);
```


i've also noted the constant usage of the `map clone $ filter (not . null) array` snippet.
we should make that into an helper function (perhaps `functionalCopy`) , but thats left to another day. code will be even more clean if this suggestion is also applied.

```java
this.sideEffects = Stream.concat(Arrays.stream(this.sideEffects), FunctionalUtils.functionalCopy(sideEffects)).toArray(SideEffect[]::new);

Comment (roridev) [reply src/main/java/cn/nukkit/event/entity/EntityDamageEvent.java:152]: gh decided to mark the wrong lines for me. the code i'm referring to is this snippet here
https://github.com/PowerNukkit/PowerNukkit/blob/d67b5c9a36103843c502d0630f1a0eb76af93483/src/main/java/cn/nukkit/event/entity/EntityDamageEvent.java#L152-L160

**Review by roridev — APPROVED** [review]

**Pull request merged**